src/backend/test/upgrade-discovery.test.js

```diff
@@ -0,0 +1,268 @@
+// @ts-check
+const test = require("tape");
+const path = require("path");
+const getPort = require("get-port");
+const { once } = require("events");
+const hasha = require("hasha");
+const { promises: fsPromises } = require("fs");
+const { setupStorage, hashCmp } = require("./helpers.js");
+const Discovery = require("../lib/upgrade-discovery");
+const Server = require("../lib/upgrade-server");
+const fakeApkInfo = require("./fixtures/fake-apk-info");
+const _ = require("lodash");
+const getStream = require("get-stream");
+
+const validApksFolder = path.join(__dirname, "./fixtures/valid-apks");
+
+// WARNING: These tests are dependent on timing. By default, available
+// installers are removed after they have not been seen for 4000ms (TTL).
+// Discovery lookup for available peers happens every 2000ms (lookupInterval),
+// and discovery will emit the available installers it has found every 5000ms
+// (emit throttle). Tests for removed peers rely on the installer emit throttle
+// being greater than the TTL (because they take a peer offline then wait for
+// the next emit of installers)
+
+/**
+ * @param {string[]} apkFilenames List of filenames to copy into temp directory
+ * @param {import("../lib/types").InstallerInt} currentApkInfo
+ * @param {{ throttle?: number, lookupInterval?: number }} [options]
+ */
+async function createDiscovery(
+  apkFilenames,
+  currentApkInfo,
+  { throttle, lookupInterval } = {}
+) {
+  const filepaths = apkFilenames.map(filename =>
+    path.join(validApksFolder, filename)
+  );
+  const { storage, expected, cleanup } = await setupStorage(
+    filepaths,
+    currentApkInfo
+  );
+  const discovery = new Discovery({
+    discoveryKey: "com.mapeo",
```

Comment:
TODO: use bundleId to determine key

Reply:
This is just in the test (hard-coded so that test instances discovery each other). In the code it used bundleId: https://github.com/digidem/mapeo-mobile/blob/p2p-update/typescript/src/backend/lib/upgrade-manager.js#L103
However, we should also test to make sure changing this means discovery does not happen.

Reply:
Added a test for changing discovery keys (to ensure they don't discovery each other)